worker/raft/rafttransport/worker.go

```diff
@@ -0,0 +1,336 @@
+// Config is the configuration required for running an apiserver-based
+// raft transport worker.
+type Config struct {
+	// APIInfo contains the information, excluding addresses,
+	// required to connect to an API server.
+	APIInfo *api.Info
+
+	// APIOpen is the function used to dial an API connection.
+	APIOpen api.OpenFunc
+
+	// Hub is the structured hub to which the worker will subscribe
+	// for API server address updates.
+	Hub *pubsub.StructuredHub
+
+	// Mux is the API server HTTP mux into which the handler will
+	// be installed.
+	Mux *apiserverhttp.Mux
+
+	// Path is the path of the raft HTTP endpoint.
+	Path string
+
+	// Tag is the tag of the agent running this worker.
+	Tag names.Tag
+
+	// Timeout, if non-zero, is the timeout to apply to transport
+	// operations. See raft.NetworkTransportConfig.Timeout for more
+	// details.
+	Timeout time.Duration
+}
+
+// Validate validates the raft worker configuration.
+func (config Config) Validate() error {
+	if config.APIInfo == nil {
+		return errors.NotValidf("nil APIInfo")
+	}
+	if config.APIOpen == nil {
+		return errors.NotValidf("nil APIOpen")
+	}
+	if config.Hub == nil {
+		return errors.NotValidf("nil Hub")
+	}
+	if config.Mux == nil {
+		return errors.NotValidf("nil Mux")
+	}
+	if config.Path == "" {
+		return errors.NotValidf("empty Path")
+	}
+	if config.Tag == nil {
+		return errors.NotValidf("nil Tag")
+	}
+	return nil
+}
+
+// NewWorker returns a new apiserver-based raft transport worker,
+// with the given configuration. The worker itself implements
+// raft.Transport.
+func NewWorker(config Config) (*Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &Worker{
+		config:         config,
+		connections:    make(chan net.Conn),
+		dialRequests:   make(chan dialRequest),
+		serversUpdated: make(chan struct{}),
+	}
+
+	const logPrefix = "[transport] "
+	logWriter := &raftutil.LoggoWriter{logger, loggo.DEBUG}
+	logLogger := log.New(logWriter, logPrefix, 0)
+	transport := raft.NewNetworkTransportWithConfig(&raft.NetworkTransportConfig{
+		Logger:                logLogger,
+		MaxPool:               7, // 7 is the maximum juju cluster size
+		ServerAddressProvider: serverAddressProvider{},
+		Stream: newStreamLayer(config.Tag, w.connections, &Dialer{
+			APIInfo: config.APIInfo,
+			DialRaw: w.dialRaw,
+			Path:    config.Path,
+		}),
+		Timeout: config.Timeout,
+	})
+	w.Transport = transport
+
+	// Subscribe to API server address changes.
+	unsubscribeHub, err := w.config.Hub.Subscribe(
+		pubsubapiserver.DetailsTopic,
+		w.apiserverDetailsChanged,
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "subscribing to apiserver details")
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: func() error {
+			defer transport.Close()
+			defer unsubscribeHub()
+			return w.loop()
+		},
+	}); err != nil {
+		transport.Close()
+		unsubscribeHub()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
```

Comment:
it took me several passes to work out where this "err != nil" was coming from (that it was from the original if).
I think this passes the usefulness for nested declarations. Can we pull out either the Plan object or the Work func so that the code is a bit more obvious here?